src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java

```diff
@@ -62,6 +86,91 @@
     return caseData;
   }
 
+  @Override
+  public CaseDTO modifyAddress(final UUID caseId, final AddressChangeDTO addressChanges)
+      throws CTPException {
+
+    Optional<CollectionCase> caseMatch = dataRepo.readCollectionCase(caseId.toString());
+
+    if (!caseMatch.isPresent()) {
+      log.with("caseId", caseId).error("Failed to retrieve Case from storage");
+      throw new CTPException(CTPException.Fault.RESOURCE_NOT_FOUND, "Failed to retrieve Case");
+    }
+
+    CollectionCase rmCase = caseMatch.get();
+
+    CaseDTO caseData = createModifiedAddressCaseDetails(caseId, rmCase, addressChanges);
+
+    AddressModified originalAddress = addressMapperFacade.map(rmCase.getAddress());
+
+    AddressModified updatedAddress =
+        addressDTOMapperFacade.map(
+            addressChanges.getAddress(), addressMapperFacade.map(rmCase.getAddress()));
+
+    sendAddressModifiedEvent(caseId, originalAddress, updatedAddress);
+
+    return caseData;
+  }
+
+  /**
+   * Create case details with updated Address
+   *
+   * @param caseId requested caseId for which to update address
+   * @param rmCase original Case from repository
+   * @param addressChanges Changed address details from request
+   * @return CaseDTO updated case details
+   * @throws CTPException UPRN of stored address and request change do not match
+   */
+  private CaseDTO createModifiedAddressCaseDetails(
+      UUID caseId, CollectionCase rmCase, AddressChangeDTO addressChanges) throws CTPException {
+
+    CaseDTO caseData = mapperFacade.map(rmCase, CaseDTO.class);
+    if (!caseData.getAddress().getUprn().equals(addressChanges.getAddress().getUprn())) {
+      log.with("caseId", caseId)
+          .with("UPRN", addressChanges.getAddress().getUprn().toString())
+          .error("Address CaseId and UPRN do not match");
+      throw new CTPException(
+          CTPException.Fault.BAD_REQUEST, "Address CaseId and UPRN do not match");
```

Comment:
"The UPRN of the referenced Case and the provided Address UPRN must be matching"

Reply:
Wording of exception changed as above.